### PR TITLE
Add support for "untied" embedding weights in language models

### PR DESCRIPTION
**Pull request by mattdangerw**
This PR adds support for language models with "untied weights"--where the weights for the output projection are separate from the weights for the input embedding. We need this for T5 and some other newer models.

To do this, we will extend our token embedding to also support a "reverse projection." We will introduce a new `ReversibleEmbedding` layer, and if `reverse=True` is passed to call, we project from `output_dim` back to `input_dim`. Handling this all in the token embedding layer allows us to have an easy setup where one layer clearly owns all the large embedding matrices.

We also need to introduce a backwards incompatible change for the `MaskedLMHead` layer, to take as input a layer instead of a shared variable. From discussion on https://github.com/keras-team/keras-core/issues/681, passing a shared variable to init is not actually a good approach; we can rely on layers being deduplicated properly, but not weights. All our masked lm tasks will work unaltered (e.g. `keras_nlp.models.BertMaskedLM`), this API change is just for the layer.

Side note, this gives us a nice recipe for converting a `backbone` to a simple language model.
```python
backbone = ...
inputs = backbone.input
hidden_states = backbone(inputs)
output = backbone.token_embedding(hidden_states, reverse=True)
language_model = keras.Model(inputs, outputs)
```

### CODE REVIEW

File: keras_nlp/layers/modeling/masked_lm_head.py [hunk 1]
```diff
@@ -46,11 +46,9 @@ class MaskedLMHead(keras.layers.Layer):
 
     Args:
         vocabulary_size: The total size of the vocabulary for predictions.
-        embedding_weights: Optional. The weights of the word embedding used
-            to transform input token ids. The transpose of this weight matrix
-            will be used to project a token embedding vector to a prediction
-            over all input words, as described
-            [here](https://arxiv.org/abs/1608.05859).
+        token_embedding: Optional. A `keras_nlp.layers.ReversibleEmbedding`. If
```
Comment (fchollet) [inline keras_nlp/layers/modeling/masked_lm_head.py:49]: Insert: "instance"

File: keras_nlp/layers/modeling/masked_lm_head.py [hunk 2]
```diff
@@ -46,11 +46,9 @@ class MaskedLMHead(keras.layers.Layer):
...
+            passed, the layer will be used to project from the hidden_dim of the
```
Comment (fchollet) [inline keras_nlp/layers/modeling/masked_lm_head.py:50]: Backticks

File: keras_nlp/layers/modeling/reversible_embedding.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import numpy as np
+
+from keras_nlp.api_export import keras_nlp_export
+from keras_nlp.backend import keras
+from keras_nlp.backend import ops
+
+
+@keras_nlp_export("keras_nlp.layers.ReversibleEmbedding")
+class ReversibleEmbedding(keras.layers.Embedding):
+    """A `keras.layers.Embedding` which can project backwards to the input dim.
```
Comment (fchollet) [inline keras_nlp/layers/modeling/reversible_embedding.py:23]: "An embedding layer"

File: keras_nlp/layers/modeling/reversible_embedding.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        embeddings_constraint: Constraint function applied to
+            the `embeddings` matrix (see `keras.constraints`).
+        mask_zero: Boolean, whether or not the input value 0 is a special
+            "padding" value that should be masked out.
+
+    Call args:
+        inputs: The tensor inputs to the layer.
+        reverse: Boolean. If `True` the layer will perform a linear projection
+            from `output_dim` to `input_dim`, instead of a normal embedding
+            call.
```
Comment (fchollet) [inline keras_nlp/layers/modeling/reversible_embedding.py:56]: "Defaults to `False`."

File: keras_nlp/layers/modeling/reversible_embedding.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+    embedding = keras.layers.Embedding(vocab_size, hidden_dim)
+    # Embed tokens to shape `(batch_size, seq_length, hidden_dim)`.
+    hidden_states = embedding(token_ids)
+    # Project hidden states to shape `(batch_size, seq_length, vocab_size)`.
+    logits = embedding(hidden_state, reverse=True)
+    ```
+
+    References:
+     - [Vaswani et al., 2017](https://arxiv.org/abs/1706.03762)
```
Comment (fchollet) [inline keras_nlp/layers/modeling/reversible_embedding.py:76]: Unindent

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. Looking good!

**Comment by abheesht17** [comment]
Name's not `TransposedEmbedding`? :(

(just kidding :P)

**Comment by abheesht17** [comment]
Name's not `TransposedEmbedding`? :(

(just kidding :P)

**Review by jbischof — APPROVED** [review]
Looks great thank you!

Unfortunately we need to update the "getting started" guide once again. I wish there were a lightweight way to do this.

File: keras_nlp/layers/modeling/reversible_embedding.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    By default, the reverse projection will use the transpose of the
+    `embeddings` weights to project to `input_dim` (weights are "tied"). If
+    `tie_weights=False`, the model will use a separate, trainable variable for
+    reverse projection.
+
+    This layer has no bias terms.
+
+    Args:
+        input_dim: Integer. Size of the vocabulary,
+            i.e. maximum integer index + 1.
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/layers/modeling/reversible_embedding.py:38]: I just copied the embedding documentation -> https://keras.io/api/layers/core_layers/embedding/

I think the intent is just to remind users that if the highest valid token id is 100, your vocab size is 101. No strong feelings on what it says, but we should probably stay 1-1 with the Keras documentation here.
